src/backend/utils/adt/misc.c

```
@@ -368,6 +369,47 @@ pg_terminate_backend_msg(PG_FUNCTION_ARGS)
 	PG_RETURN_BOOL(r == SIGNAL_BACKEND_SUCCESS);
 }
 
+Datum
+gp_terminate_mpp_backends(PG_FUNCTION_ARGS)
+{
+	int count;
+	const int retry_count = 15;
+
+	if (Gp_role != GP_ROLE_EXECUTE)
+		ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+						(errmsg("Only terminate mpp backends on segments"))));
+
+	if (!superuser())
+		ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+						(errmsg("Superuser only to execute it."))));
+
+	elog(LOG, "Terminating all mpp backends except myself");
+
+	count = 0;
+	while (SignalMppBackends(SIGTERM) > 0 && count < retry_count)
+	{
+		CHECK_FOR_INTERRUPTS();
+		pg_usleep(1000 * 1000L); /* wait for 1s */
+		count++;
+	}
+
+	if (count == retry_count)
+	{
+		count = 0;
+		while (SignalMppBackends(SIGQUIT) > 0 && count < retry_count)
```

The exit code of a backend is 2 when it receives SIGQUIT. So, the backend of the UDF will also be killed, and the QD has to ignore the errors.
I prefer the segment kills their MPP backends rather than dispatching the killing command to the segment.

---

No worry since SignalMppBackends() skips itself when sending signals. What do you mean "the segment kills their mpp backends"?

---

The signal handler for SIGQUIT is `quickdie()`, it exits the process with exit code 2. So it will cause the postmaster to reset the whole instance, including killing the process that calling SignalMppBackends(). Isn't it?

---

Yes that seems to be a concern so we may finally remove the SIGQUIT code, just use SIGTERM similar to pg_terminate_backend(). or maybe just trigger a reset if SIGTERM fails finally. Let me further think a bit. Thanks.

---

I'm now using elog(PANIC) if SIGTERM does not resolve this issue after timeout.

---

But I called setrlimit() to avoid potential core file generation.